### PR TITLE
New feature docs

### PR DESCRIPTION
**Pull request by greg-mckeon**
There's more to add here, but starting the PR off.

### CODE REVIEW

File: src/content/learning/how-durable-objects-work.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+There is no support for querying, bulk loading, or exporting data stored in Durable Objects.
+
+Applications should be prepared for transactions to fail due to conflicts, and to retry those failed transactions.  A failed transaction will throw an error (TODO: what error?  is this true?) that can be caught.  It is always safe to retry failed transactions.
+
+### Performance
+
+Durable Objects are single-threaded, but each Object has almost no overhead associated with it.  In general, this means your application should look to create and access many Durable Objects over sending many requests to the same Object.  Access patterns that make many requests to a single Object will see degraded performance.
+
+Since Durable Objects are globally distributed, initial creation latency may be high when using system-generated ids (see above).  Until automatic migration of Durable Objects is implemented, latency will also be high on subsequent read and write operations.
```
Comment (koeninger) [inline src/content/learning/how-durable-objects-work.md]: Is this intended to say client-defined ids?

File: src/content/learning/how-durable-objects-work.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+Learn more at the [Workers Durable Objects API reference](/runtime-apis/durable-objects).
+
+## Limitations and considerations when working with Durable Objects
+
+### Global Uniqueness
+
+Uniqueness is currently enforced only when a Durable Object begins execution.  In the future, uniqueness will also be enforced whenever a Durable Object makes storage requests.  
+
+During a network partition between data centers or between machines within a data center, it is possible for more than one Durable Object with a given id to be active.  While this is unlikely, we plan to do additional work to reduce the chance of a network partition during the beta period.
```
Comment (koeninger) [inline src/content/learning/how-durable-objects-work.md]: What are we planning to do to reduce the chance of a network partition?

File: src/content/learning/how-durable-objects-work.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+```
+
+Learn more at the [Workers Durable Objects API reference](/runtime-apis/durable-objects).
+
+## Limitations and considerations when working with Durable Objects
+
+### Global Uniqueness
+
+Uniqueness is currently enforced only when a Durable Object begins execution.  In the future, uniqueness will also be enforced whenever a Durable Object makes storage requests.
```
Comment (koeninger) [inline src/content/learning/how-durable-objects-work.md]: I'd explicitly call out that this means when a script is updated, existing requests may still be running on the old object concurrently with new requests going to the new object.

File: src/content/learning/using-durable-objects.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+order: 8
+---
+
+# Using Durable Objects
+
+Durable Objects provide low-latency coordination and consistent storage for the Workers platform through two features: global uniqueness and a transactional storage API.
+
+* Global Uniqueness guarantees that there will be a single Durable Object with a given id running at once.  Requests for a Durable Object id are routed by the Workers runtime to the Cloudflare point-of-presence that owns the Durable Object.  
+
+* The transactional storage API provides strongly-consistent, key-value storage to the Durable Object.  Each Object can only read and modify keys associated with that Object. Execution of a Durable Object is single-threaded, but multiple request threads may be processed out-of-order from how they arrived at the Object.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: > but multiple request threads may be processed out-of-order from how they arrived at the Object.

The wording is here is a little strange. Maybe switch it to "multiple request events" instead of threads

File: src/content/learning/using-durable-objects.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+* __Writing the class__ that defines a Durable Object.
+* __Binding that class__ into a Worker.
+* __Instantiating a Durable Object__ from within a running Worker.
+* __Communicating with a Durable Object__ from a running Worker via the Fetch API.
+
+## Writing a class that defines a Durable Object
+
+To define a Durable Object, developers export an ordinary Javascript class.  Other languages will need a shim that translates their class definition to a Javascript class.
+
+The first variable passed to the class constructor contains state specific to the Durable Object, including methods for accessing storage.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: The natural next question that doesn't appear to be answered here is what's the second variable passed to the constructor?

File: src/content/learning/using-durable-objects.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+Workers communicate with a Durable Object via the fetch API.  Like any other Worker, a Durable Object listens for incoming Fetch events by registering an event handler.
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+    }
+    
+    addEventListener("fetch", (event) => {
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: `addEventListener` isn't used in class-based actors. There just needs to be a function with the prototype `async fetch(request) {`

File: src/content/learning/using-durable-objects.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+}
+
+```
+
+A Worker can pass information to a Durable Object via headers, the HTTP method, the Request body, or the Request URI.
+
+### Accessing Storage from a Durable Object
+
+Durable Objects gain access to a TODO: fill in[persistent storage API]() via the first parameter passed to the Durable Object constructor.  While access to a Durable Object is single-threaded, it's important to remember that there can still be race conditions across multiple requests.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: > it's important to remember that there can still be race conditions across multiple requests.

I'd reword this to something like "it's important to remember that request executions can still interleave with each other when they wait on I/O, such as when waiting on the promises returned by persistent storage methods or `fetch` requests."

File: src/content/learning/using-durable-objects.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+        this.state = state
+    }
+    
+    addEventListener("fetch", (event) => {
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: Again, get rid of this and change the name of the function below from `handle` to `fetch`.

File: src/content/learning/using-durable-objects.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+        let storagePromise = this.state.storage.set(ip, data);
+        await storagePromise;
+        return new Response(ip + ' stored ' + data);
+    }
+
+}
+
+```
+
+Single statement transactions are always transactional.  More complex use cases can wrap multiple storage statements in a transaction.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: "Single statement transactions are always transactional" comes off sounding a bit strange. Perhaps "Each individual storage operation behaves like a database transaction."

File: src/content/learning/using-durable-objects.md [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+Single statement transactions are always transactional.  More complex use cases can wrap multiple storage statements in a transaction.  
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+        this.state = state
+    }
+    
+    addEventListener("fetch", (event) => {
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: Again, get rid of this and change the name of the function below from `handle` to `fetch`.

File: src/content/learning/using-durable-objects.md [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+
+}
+
+```
+
+Single statement transactions are always transactional.  More complex use cases can wrap multiple storage statements in a transaction.  
+
+```js
+
+class DurableObjectExample {
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: It'd be nice to name each of these class differently for each example as a way of indicating what the example is meant to show.

File: src/content/learning/using-durable-objects.md [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+ 
+    async function handle(request) {
+        // TODO: txn example
+    }
+
+}
+
+```
+
+Transactions can fail if they conflict with a concurrent transaction from the same Durable Object.  Transactions are transparently and automatically retried once before returning a failure.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: I'd insert an extra little clause here, changing "retried once before" to "retried once by rerunning the provided function before"

Comment (a-robinson) [reply src/content/learning/using-durable-objects.md]: This could also be a place to add in that this is because all transactions are run at serializable isolation, making them strongly consistent. Totally optional though

File: src/content/learning/using-durable-objects.md [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+
+
+## Binding the class to a Worker
+
+As with Workers KV, the class namespace must be created and then bound into a Worker before it can be used.
+
+First, upload the script that contains the class via the Workers Dashboard or via the command line.
+
+```sh
+curl -i -H "X-Auth-Email: ${EMAIL}" -H "X-Auth-Key: ${AUTH_KEY}" "https://api.cloudflare.com/client/v4/accounts/${ACCOUNT_TAG}/workers/scripts/${SCRIPT_NAME}" -X PUT -H "Content-Type: application/javascript" --data "`cat durable_object_example.js`"
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: We might encourage people to use API tokens rather than auth keys, which just requires a header like "Authorization: Bearer <token>". Probably worth checking out what other parts of the docs use.

File: src/content/learning/using-durable-objects.md [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+```
+
+To create the class, call the following endpoint.
+
+```sh
+curl -i -H "X-Auth-Email: ${EMAIL}" -H "X-Auth-Key: ${AUTH_KEY}" "https://api.cloudflare.com/client/v4/accounts/${ACCOUNT_TAG}/workers/durable_objects/namespaces" -X POST --data "{'name': 'example-class', 'script': ${SCRIPT_NAME}"
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: The JSON blob appears syntactically invalid, and also a class name needs to be specified  as well. Use this instead:

```
"{'name': 'example-class', 'script': 'example-script', 'class: 'example-class'}"
```

File: src/content/learning/using-durable-objects.md [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+You can now create a binding for the Worker which uses this class.  TODO: Not sure if this is still the correct syntax, check.
+
+```js 
+// binding.json
+{
+  "body_part": "script",
+  "bindings": [
+    {
+      "type": "durable_object_namespace",
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md:243]: Needs to be "durable_object_class"

File: src/content/learning/using-durable-objects.md [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+      "name": "example-class",
+      "namespace_id": "<namespace id>"
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: Needs to be "class_id" instead of "namespace_id"

File: src/content/learning/using-durable-objects.md [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+
+```js
+
+addEventListener('fetch', event => {
+  event.respondWith(handleRequest(event.request))
+})
+ 
+async function handleRequest(request) {
+  // Retrieve the Fetcher for the Durable Object with id 1
+  let durableObject = await DURABLE_OBJECT_CLASS.get("1")
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: It'd be nice to make this binding name ("DURABLE_OBJECT_CLASS") match the "name" provided in the binding definition in the section above, since that's where it comes from.

Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: This isn't how requesting an actor works. You first need to get an ID object, then pass it to the durable object class binding, like this:

```
let objectId = DURABLE_OBJECT_CLASS.idFromName("1")
let durableObject = await DURABLE_OBJECT_CLASS.get(objectId)
```

File: src/content/learning/using-durable-objects.md [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+  // Send a request to Durable Object '1', with the URI "get-count"
+  let resp = await durableObject.fetch("get-count")
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: All URIs need to be validly formed, including either an `http://` or `https://` prefix. We'll likely remove this restriction in the future, but it's important for now. So change this to something like `"https://get-count"`

File: src/content/learning/using-durable-objects.md [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+  let count = await resp.text()
+ 
+  // Regardless of the URI, requests are always routed to the Durable Object
+  resp = await durableObject.fetch()
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: A fetch without a URL at all definitely won't work.

Before publishing we'll need to make sure to run these examples for real and make sure they all work.

File: src/content/learning/using-durable-objects.md [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+  let message = await resp.text()
+
+  // TODO: system generated id example
+  let id = generateId()
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]:
```
  let systemGeneratedId = DURABLE_OBJECT_CLASS.newUniqueId()
  let sysDurableObject = await DURABLE_OBJECT_CLASS.get(systemGeneratedId)
  resp = await sysDurableObject.fetch('https://get-count')
```

File: src/content/learning/using-durable-objects.md [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+
+```
+
+Learn more at the [Workers Durable Objects API reference](/runtime-apis/durable-objects).
+
+## Limitations and considerations when working with Durable Objects
+
+### Global Uniqueness
+
+Uniqueness is currently enforced only when a Durable Object begins execution.  In the future, uniqueness will also be enforced whenever a Durable Object makes storage requests.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: > In the future, uniqueness will also be enforced whenever a Durable Object makes storage requests.  

If we want to be more specific, at the present this only time during which you can currently have overlap of an old instance of a durable object and a new instance is when pushing a new version of your code. It's understandable if we don't want to be that specific though.

Along those lines, though, it's worth including an example that shows using the class's in-memory state for something, and then perhaps calling out down here that in-memory state is transient and can disappear when an actor needs to be stopped or replaced, such as when a new its code is pushed or when it doesn't receive traffic for some time between requests.

File: src/content/learning/using-durable-objects.md [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+
+Learn more at the [Workers Durable Objects API reference](/runtime-apis/durable-objects).
+
+## Limitations and considerations when working with Durable Objects
+
+### Global Uniqueness
+
+Uniqueness is currently enforced only when a Durable Object begins execution.  In the future, uniqueness will also be enforced whenever a Durable Object makes storage requests.  
+
+During a network partition between data centers or between machines within a data center, it is possible for more than one Durable Object with a given id to be active.  While this is unlikely, we plan to do additional work to reduce the chance of a network partition during the beta period.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: This is not true. The only sense in which it's possible for multiple instances of a durable object to be active at the same is that requests to an old instance might still be executing when the new instance starts receiving requests. It is explicitly **not** possible for a partition to cause two different instances of an object to be receiving requests at the same point in time.

File: src/content/learning/using-durable-objects.md [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+
+### Creation and Routing
+
+Durable Objects use a deterministic hash of the Object's id to determine the Cloudflare point-of-presence that owns the Object.  It is not currently possible for an Object to migrate ownership to a different point-of-presence.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: > Durable Objects use a deterministic hash of the Object's id to determine the Cloudflare point-of-presence that owns the Object.

We use what's effectively a hash for durable object IDs generated using `ns.idFromName('some-string')`. We always put durable objects created using a system-generated ID in the closest possible point-of-presence. I see that your next sentence covers system generated IDs, but this sentence should not exist, at least not without making clear that it's only true for client-defined IDs.

Honestly, I'd just remove these two sentences, and put the sentence about how objects can not currently be migrated to different parts of the world before what's currently the last sentence in this section.

File: src/content/learning/using-durable-objects.md [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+
+When using system generated ids, Durable Objects are created in a Cloudflare point-of-presence located near the Worker that made the creation request.
+
+When using client-defined ids, Durable Objects may be created in any of Cloudflare's points-of-presence. Durable Object with client-defined ids will see greater variance in request latency across different Objects, as each Object may be routed to a different point-of-presence.  
+
+When migrations of Objects are enabled, this variance in request latency will only affect Object creation.
+
+### Transactional Storage API
+
+The transactional storage API is scoped to a single Durable Object.  It is not currently possible to access data stored in a Durable Object from a different Durable Object directly through the API.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: > It is not currently possible to access data stored in a Durable Object from a different Durable Object directly through the API

Minor nit, but I'd reword this a little to be:
"It is not currently possible to access data stored in one Durable Object from a different Durable Object."

File: src/content/learning/using-durable-objects.md [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+
+There is no support for querying, bulk loading, or exporting data stored in Durable Objects.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: "querying" is such a vague word, and it's not that it isn't possible to load or export data so much as that you have to write your own code to do it for now. If you don't mind all the edits, I'd make this sentence:

"There is no support for advanced queries or bulk loading or exporting of data. Loading or exporting data can be done by writing custom Workers and Durable Objects to do so yourself.

File: src/content/learning/using-durable-objects.md [hunk 26]
```diff
@@ -0,0 +1,216 @@
...
+
+Applications should be prepared for transactions to fail due to conflicts, and to retry those failed transactions.  A failed transaction will throw an error (TODO: what error?  is this true?) that can be caught.  It is always safe to retry failed transactions.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: I'd say "Applications that use multi-operation transactions ..." to be clear.

Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: The error thrown will include text saying "Transaction failed due to conflict".

File: src/content/learning/using-durable-objects.md [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+
+### Performance
+
+Durable Objects are single-threaded, but each Object has almost no overhead associated with it.  In general, this means your application should look to create and access many Durable Objects over sending many requests to the same Object.  Access patterns that make many requests to a single Object will see degraded performance.
+
+Since Durable Objects are globally distributed, initial creation latency may be high when using user-generated ids (see above).  Until automatic migration of Durable Objects is implemented, latency will also be high on subsequent read and write operations.
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: s/user-generated/user-provided

Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: > latency will also be high on subsequent read and write operations.

"latency will also be high on subsequent requests to such Durable Objects".

**Review by a-robinson — COMMENTED** [review]
Sorry for the big dump of comments. Are you gonna need some help finishing these up?

File: src/content/learning/how-durable-objects-work.md (hunk 2 above)
Comment (a-robinson) [reply src/content/learning/how-durable-objects-work.md]: Ignore this, network partitions can't cause multiple durable objects with the same ID to be receiving requests at the same time. This adds nothing to the previous paragraph other than confusion.

**Comment by a-robinson** [comment]
Pushed updates to resolve my remaining comments that Kenton's changes didn't already handle.

**Comment by kentonv** [comment]
Pushed some more stuff. I realize after the fact that I didn't follow the usual pattern for documenting class methods when I documented bindings, but I'm out of time.

I started calling the bindings "namespaces" because calling them "class bindings" was getting too confusing. I kind of think we may want to change the API back to `/workers/durable_objects/namespaces` rather than `/classes`... you configure a namespace that uses a particular class. Calling the namespace itself a "class" leads to too much confusion when trying to explain them.

Another, wilder idea: Maybe we shouldn't make people configure the namespaces separately at all. Maybe they could be implicitly created when you upload the script. (This idea obviously has lots of issues to work through but wanted to plant the thought to chew on...)

File: src/content/learning/how-durable-objects-work.md (hunk 2 above)
Comment (greg-mckeon) [reply src/content/learning/how-durable-objects-work.md]: For my own understanding, this is because the a durable object will need to write to the ownership table, which it won't be able to do if there's a partition?

**Comment by greg-mckeon** [comment]
Not merging just yet, want to give @koeninger a chance to look.  Then will add a separate PR for documenting the API specifically.

File: src/content/learning/using-durable-objects.md [hunk 28]
```diff
@@ -0,0 +1,302 @@
...
+
+<Aside header="String-derived IDs vs. system-generated IDs">
+
+In the above example, we used a string-derived object ID. You can also ask the system generate random unique IDs. System-generated unique IDs have better performance characteristics, but require that you store the ID somewhere in order to access the object again later. [See the API reference docs for more information.](/runtime-apis/durable-objects)
+
+</Aside>
+
+## Limitations during the Beta
+
+<<<<<<< HEAD
```
Comment (koeninger) [inline src/content/learning/using-durable-objects.md]: Just want to make sure this conflict is handled before merging

**Comment by koeninger** [comment]
I'm still in favor of just linking to the shell script rather than all of the details about the upload api.

**Comment by koeninger** [comment]
Not a big deal, but I think it should be possible to add the shell script as a static file and link it rather than copy-paste, e.g. https://www.gatsbyjs.com/docs/static-folder/

But this seems fine for now.

File: src/content/learning/using-durable-objects.md [hunk 29]
```diff
@@ -0,0 +1,369 @@
...
+
+        // Return current value.
+        return new Response(this.value);
+    }
+}
+```
+
+## Configuration Script
+
+To simplify steps
```
Comment (a-robinson) [inline src/content/learning/using-durable-objects.md]: Left this sentence hanging? I assume it's meant to link to the script?

**Comment by greg-mckeon** [comment]
Add Durable Objects docs
